srtp.go

```diff
@@ -54,45 +54,60 @@ func (c *Context) DecryptRTP(dst, encrypted []byte, header *rtp.Header) ([]byte,
 	return c.decryptRTP(dst, encrypted, header)
 }
 
-// EncryptRTP encrypts a plaintext RTP packet, writing to the dst buffer provided.
-// If the dst buffer does not have the capacity to hold `len(plaintext) + 10` bytes, a new one will be allocated.
-// If a rtp.Header is provided, it will be Unmarshaled using the plaintext.
-func (c *Context) EncryptRTP(dst []byte, plaintext []byte, header *rtp.Header) ([]byte, error) {
-	// TODO(@lcurley) Potentially accept a *rtp.Packet to avoid this Unmarshal
-	if header == nil {
-		header = &rtp.Header{}
-	}
+// EncryptRTP marshals and encrypts an RTP packet, writing to the dst buffer provided.
+// If the dst buffer does not have the capacity to hold `len(plaintext) + 10` bytes, a new one will be allocated and returned.
+// If the header is provided, it will be unmarshaled using the plaintext.
+func (c *Context) EncryptRTP(dst []byte, plaintext []byte, header *rtp.Header) (ciphertext []byte, err error) {
+	packet := &rtp.Packet{}
 
-	err := header.Unmarshal(plaintext)
+	err = packet.Header.Unmarshal(plaintext)
 	if err != nil {
 		return nil, err
 	}
 
+	packet.Payload = plaintext[packet.Header.PayloadOffset:]
+
+	// If the header is provided, populate it with our unmarshaled data.
+	if header != nil {
+		*header = packet.Header
+	}
+
+	return c.encryptRTP(dst, packet)
+}
+
+// encryptRTP marshals and encrypts an RTP packet, writing to the dst buffer provided.
+// If the dst buffer does not have the capacity to hold `len(plaintext) + 10` bytes, a new one will be allocated and returned.
+// This is similar but slightly faster than the above method.
+func (c *Context) encryptRTP(dst []byte, packet *rtp.Packet) (ciphertext []byte, err error) {
 	// Write to dst starting at this offset.
 	offset := 0
 
 	// Grow the given buffer to fit the output.
 	// authTag = 10 bytes
-	dst = growBufferSize(dst, len(plaintext)+10)
+	dst = growBufferSize(dst, packet.MarshalSize()+10)
 
-	s := c.getSSRCState(header.SSRC)
-	c.updateRolloverCount(header.SequenceNumber, s)
+	s := c.getSSRCState(packet.Header.SSRC)
+	c.updateRolloverCount(packet.Header.SequenceNumber, s)
 
 	// Copy the header unencrypted.
-	headerSize := header.PayloadOffset
+	// The dst slicing is weird because MarshalTo uses append while others write directly.
+	rawHeader, err := packet.Header.MarshalTo(dst[offset:offset])
```

Comment:
I'm actually with changing `MarshalTo` now that I've tried to use it a bit. The append functionality is okay but it doesn't match `io.Reader` like most things pions uses.

The one caveat is that it should error if `len(data) < packet.MarshalSize()`. The old code would append and accidentally create copies. It also means the API would be:

`packet.MarshalTo(data []byte) (n int, err error)` aka `io.Reader`